src/server/pkg/storage/fileset/compaction.go

```diff
@@ -0,0 +1,168 @@
+package fileset
+
+import (
+	"context"
+	"time"
+
+	"github.com/pachyderm/pachyderm/src/client/pkg/errors"
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset/index"
+	"golang.org/x/sync/errgroup"
+)
+
+// IsCompacted returns true if the fileset is already in compacted form.
+func (s *Storage) IsCompacted(ctx context.Context, id ID) (bool, error) {
+	md, err := s.store.Get(ctx, id)
+	if err != nil {
+		return false, err
+	}
+	switch x := md.Value.(type) {
+	case *Metadata_Composite:
+		ids, err := s.Flatten(ctx, x.Composite.Layers)
+		if err != nil {
+			return false, err
+		}
+		layers, err := s.getPrimitiveBatch(ctx, ids)
+		if err != nil {
+			return false, err
+		}
+		return isCompacted(s.levelFactor, layers), nil
+	case *Metadata_Primitive:
+		return true, nil
+	default:
+		// TODO: should it be?
+		return false, errors.Errorf("IsCompacted is not defined for empty filesets")
+	}
+}
+
+func (s *Storage) getPrimitiveBatch(ctx context.Context, ids []ID) ([]*Primitive, error) {
+	var layers []*Primitive
+	for _, id := range ids {
+		prim, err := s.getPrimitive(ctx, id)
+		if err != nil {
+			return nil, err
+		}
+		layers = append(layers, prim)
+	}
+	return layers, nil
+}
+
+func isCompacted(factor int64, layers []*Primitive) bool {
+	return indexOfCompacted(factor, layers) == len(layers)
+}
+
+// indexOfCompacted returns the last value of i for which the "compacted relationship"
+// is maintained for all layers[:i+1]
+// the "compacted relationship" is defined as leftSize >= (rightSize * factor)
+// If there is an element at i+1. It will be the first element which does not satisfy
+// the compacted relationship with i.
+func indexOfCompacted(factor int64, inputs []*Primitive) int {
```

Comment:
A concern I brought up about this when you initially proposed it was the duplicate work that can occur when you hit the situation where a small change results with compacting multiple of the larger levels (with this algorithm we would recursively compact them vs compacting them in one pass like the current algorithm). I think the simplicity we get from this outweighs that concern for me, and the extra work is the sum of the exponential decrease of the necessary work in the worst case (I'm not super strong at math, but I think the additional overhead is not going to be that big of a deal). Also, I think we could add the constant compaction delay (don't compact until > x number of primitive filesets) by just wrapping this with a check that counts the primitives before calling into this functionality.

Reply:
I'm going to make a ticket for Compaction Optimizations and add it to the 2.0.0 alpha project.  And keep it out of this P.R.